MA-272 video_analytics

BenjiLee opened the pull request:
@nasthagiri @mulby 

FYI @stroilova

nasthagiri commented:
:+1: LGTM.  
@mulby please review

mulby commented:
The payloads are identical between the old and new events?

nasthagiri commented:
@mulby By old and new, do you mean browser-generated and mobile-generated?

mulby commented:
yes

nasthagiri commented:
@BenjiLee will be updating the ticket with the events found in the tracking log: https://openedx.atlassian.net/browse/MA-272

BenjiLee commented:
Android
`"event": "{\"code\": \"mobile\", \"new_time\": 104.268, \"old_time\": 134.268, \"type\": \"onSkipSeek\", \"id\": \"i4x-MITx-15_390_1x-video-174238d48fd547269b05a7d0dcaa8394\", \"requested_skip_interval\": -30
"event": "{\"code\": \"mobile\", \"new_time\": 198.247, \"old_time\": 104.571, \"type\": \"onSlideSeek\", \"id\": \"i4x-MITx-15_390_1x-video-174238d48fd547269b05a7d0dcaa8394\", \"requested_skip_interval\": 93.676}", "event_type": "seek_video"`

iOS:
`"event": "{\"code\": \"mobile\", \"new_time\": 89.699177437, \"old_time\": 119.699177437, \"type\": \"onSkipSeek\", \"id\": \"i4x-MITx-15_390_1x-video-3e9b631fddb54c72b1bce30ca8643f48\", \"requested_skip_interval\": -30}", 

"event": "{\"code\": \"mobile\", \"new_time\": 93, \"old_time\": 129.138042608, \"type\": \"onSlideSeek\", \"id\": \"i4x-MITx-15_390_1x-video-3e9b631fddb54c72b1bce30ca8643f48\", \"requested_skip_interval\": -36.13804260800001
`

Browser
`"event": "{\"id\":\"i4x-MITx-15_390_1x-video-3e9b631fddb54c72b1bce30ca8643f48\",\"old_time\":1.948855,\"new_time\":45,\"type\":\"onSlideSeek\",\"code\":\"mNUpUgY9ivo\
`

BenjiLee commented:
iOS
`{
    "username": "staff",
    "event_source": "mobile",
    "name": "edx.video.position.changed",
    "accept_language": "",
    "time": "2015-03-04T18:42:59.741000+00:00",
    "agent": "",
    "page": "http://mobile3.m.sandbox.edx.org/courses/MITx/15.390.1x/3T2014/courseware/382d25ff3a224b079a399d64c6c6a3eb/4c3f30318406482c990f920c19df1f8d",
    "host": "mobile3.m.sandbox.edx.org",
    "session": "",
    "referer": "",
    "context": {
        "component": "videoplayer",
        "received_at": "2015-03-04T18:43:08.741000+00:00",
        "course_id": "MITx/15.390.1x/3T2014",
        "path": "/segmentio/event",
        "user_id": 4,
        "org_id": "MITx",
        "application": {
            "version": "1.0.01",
            "name": "edx.mobileapp.iOS"
        },
        "client": {
            "network": {
                "carrier": "T-Mobile",
                "bluetooth": true
            },
            "locale": "en-US",
            "ip": "18.111.117.111",
            "app": {
                "version": "1.0.01",
                "namespace": "org.edx.mobile",
                "build": "1.0.07",
                "name": "edX"
            },
            "library": {
                "version": "1.8.15",
                "name": "analytics-ios"
            },
            "device": {
                "advertisingId": "023321BF-BA40-43A0-ACA9-05D208A1B2E0",
                "model": "iPhone7,1",
                "type": "ios",
                "id": "6201488D-10D1-4A42-8DAB-42633729BD55",
                "adTrackingEnabled": true,
                "manufacturer": "Apple"
            },
            "timezone": "America/New_York",
            "os": {
                "version": "8.0",
                "name": "iPhone OS"
            },
            "screen": {
                "width": 320,
                "height": 568
            }
        }
    },
    "ip": "",
    "event": "{\"code\": \"mobile\", \"new_time\": 89.699177437, \"old_time\": 119.699177437, \"type\": \"onSkipSeek\", \"id\": \"i4x-MITx-15_390_1x-video-3e9b631fddb54c72b1bce30ca8643f48\", \"requested_skip_interval\": -30}",
    "event_type": "seek_video"
}{
    "username": "staff",
    "event_source": "mobile",
    "name": "edx.video.position.changed",
    "accept_language": "",
    "time": "2015-03-04T18:43:37.159000+00:00",
    "agent": "",
    "page": "http://mobile3.m.sandbox.edx.org/courses/MITx/15.390.1x/3T2014/courseware/382d25ff3a224b079a399d64c6c6a3eb/4c3f30318406482c990f920c19df1f8d",
    "host": "mobile3.m.sandbox.edx.org",
    "session": "",
    "referer": "",
    "context": {
        "component": "videoplayer",
        "received_at": "2015-03-04T18:43:42.159000+00:00",
        "course_id": "MITx/15.390.1x/3T2014",
        "path": "/segmentio/event",
        "user_id": 4,
        "org_id": "MITx",
        "application": {
            "version": "1.0.01",
            "name": "edx.mobileapp.iOS"
        },
        "client": {
            "network": {
                "carrier": "T-Mobile",
                "bluetooth": true
            },
            "locale": "en-US",
            "ip": "18.111.117.111",
            "app": {
                "version": "1.0.01",
                "namespace": "org.edx.mobile",
                "build": "1.0.07",
                "name": "edX"
            },
            "library": {
                "version": "1.8.15",
                "name": "analytics-ios"
            },
            "device": {
                "advertisingId": "023321BF-BA40-43A0-ACA9-05D208A1B2E0",
                "model": "iPhone7,1",
                "type": "ios",
                "id": "6201488D-10D1-4A42-8DAB-42633729BD55",
                "adTrackingEnabled": true,
                "manufacturer": "Apple"
            },
            "timezone": "America/New_York",
            "os": {
                "version": "8.0",
                "name": "iPhone OS"
            },
            "screen": {
                "width": 320,
                "height": 568
            }
        }
    },
    "ip": "",
    "event": "{\"code\": \"mobile\", \"new_time\": 93, \"old_time\": 129.138042608, \"type\": \"onSlideSeek\", \"id\": \"i4x-MITx-15_390_1x-video-3e9b631fddb54c72b1bce30ca8643f48\", \"requested_skip_interval\": -36.13804260800001}",
    "event_type": "seek_video"
}1

BenjiLee commented:
```json
{
    "username": "staff",
    "event_source": "mobile",
    "name": "edx.video.position.changed",
    "accept_language": "",
    "time": "2015-03-04T18:42:59.741000+00:00",
    "agent": "",
    "page": "http://mobile3.m.sandbox.edx.org/courses/MITx/15.390.1x/3T2014/courseware/382d25ff3a224b079a399d64c6c6a3eb/4c3f30318406482c990f920c19df1f8d",
    "host": "mobile3.m.sandbox.edx.org",
    "session": "",
    "referer": "",
    "context": {
        "component": "videoplayer",
        "received_at": "2015-03-04T18:43:08.741000+00:00",
        "course_id": "MITx/15.390.1x/3T2014",
        "path": "/segmentio/event",
        "user_id": 4,
        "org_id": "MITx",
        "application": {
            "version": "1.0.01",
            "name": "edx.mobileapp.iOS"
        },
        "client": {
            "network": {
                "carrier": "T-Mobile",
                "bluetooth": true
            },
            "locale": "en-US",
            "ip": "18.111.117.111",
            "app": {
                "version": "1.0.01",
                "namespace": "org.edx.mobile",
                "build": "1.0.07",
                "name": "edX"
            },
            "library": {
                "version": "1.8.15",
                "name": "analytics-ios"
            },
            "device": {
                "advertisingId": "023321BF-BA40-43A0-ACA9-05D208A1B2E0",
                "model": "iPhone7,1",
                "type": "ios",
                "id": "6201488D-10D1-4A42-8DAB-42633729BD55",
                "adTrackingEnabled": true,
                "manufacturer": "Apple"
            },
            "timezone": "America/New_York",
            "os": {
                "version": "8.0",
                "name": "iPhone OS"
            },
            "screen": {
                "width": 320,
                "height": 568
            }
        }
    },
    "ip": "",
    "event": "{\"code\": \"mobile\", \"new_time\": 89.699177437, \"old_time\": 119.699177437, \"type\": \"onSkipSeek\", \"id\": \"i4x-MITx-15_390_1x-video-3e9b631fddb54c72b1bce30ca8643f48\", \"requested_skip_interval\": -30}",
    "event_type": "seek_video"
}
{
    "username": "staff",
    "event_source": "mobile",
    "name": "edx.video.position.changed",
    "accept_language": "",
    "time": "2015-03-04T18:43:37.159000+00:00",
    "agent": "",
    "page": "http://mobile3.m.sandbox.edx.org/courses/MITx/15.390.1x/3T2014/courseware/382d25ff3a224b079a399d64c6c6a3eb/4c3f30318406482c990f920c19df1f8d",
    "host": "mobile3.m.sandbox.edx.org",
    "session": "",
    "referer": "",
    "context": {
        "component": "videoplayer",
        "received_at": "2015-03-04T18:43:42.159000+00:00",
        "course_id": "MITx/15.390.1x/3T2014",
        "path": "/segmentio/event",
        "user_id": 4,
        "org_id": "MITx",
        "application": {
            "version": "1.0.01",
            "name": "edx.mobileapp.iOS"
        },
        "client": {
            "network": {
                "carrier": "T-Mobile",
                "bluetooth": true
            },
            "locale": "en-US",
            "ip": "18.111.117.111",
            "app": {
                "version": "1.0.01",
                "namespace": "org.edx.mobile",
                "build": "1.0.07",
                "name": "edX"
            },
            "library": {
                "version": "1.8.15",
                "name": "analytics-ios"
            },
            "device": {
                "advertisingId": "023321BF-BA40-43A0-ACA9-05D208A1B2E0",
                "model": "iPhone7,1",
                "type": "ios",
                "id": "6201488D-10D1-4A42-8DAB-42633729BD55",
                "adTrackingEnabled": true,
                "manufacturer": "Apple"
            },
            "timezone": "America/New_York",
            "os": {
                "version": "8.0",
                "name": "iPhone OS"
            },
            "screen": {
                "width": 320,
                "height": 568
            }
        }
    },
    "ip": "",
    "event": "{\"code\": \"mobile\", \"new_time\": 93, \"old_time\": 129.138042608, \"type\": \"onSlideSeek\", \"id\": \"i4x-MITx-15_390_1x-video-3e9b631fddb54c72b1bce30ca8643f48\", \"requested_skip_interval\": -36.13804260800001}",
    "event_type": "seek_video"
}```

BenjiLee commented:
```json
{
    "username": "staff",
    "event_source": "mobile",
    "name": "edx.video.position.changed",
    "accept_language": "",
    "time": "2015-03-04T18:40:23.981000+00:00",
    "agent": "Dalvik/1.6.0 (Linux; U; Android 4.4.4; SM-G900V Build/KTU84P)",
    "page": "http://mobile3.m.sandbox.edx.org/courses/MITx/15.390.1x/3T2014/courseware/1a449d2fa9c8414697d550e53f387d0c/9d0771e62a3f457bb6d5ca131bc6f921",
    "host": "mobile3.m.sandbox.edx.org",
    "session": "",
    "referer": "",
    "context": {
        "component": "videoplayer",
        "received_at": "2015-03-04T18:40:23.981000+00:00",
        "course_id": "MITx/15.390.1x/3T2014",
        "path": "/segmentio/event",
        "user_id": 4,
        "org_id": "MITx",
        "application": {
            "version": "1.0.04",
            "name": "edx.mobileapp.android"
        },
        "client": {
            "network": {
                "wifi": true,
                "carrier": "Verizon Wireless",
                "cellular": true,
                "bluetooth": false
            },
            "locale": "en-US",
            "app": {
                "name": "edX",
                "packageName": "org.edx.mobile",
                "version": "1.0.04",
                "build": "org.edx.mobile@46",
                "versionName": "1.0.04",
                "versionCode": 46
            },
            "library": {
                "version": 247,
                "name": "analytics-android",
                "versionName": "2.4.7-SNAPSHOT"
            },
            "timezone": "America/New_York",
            "device": {
                "name": "kltevzw",
                "advertisingId": "d38b1986-5346-4321-994c-e0700ab76c39",
                "model": "SM-G900V",
                "type": "android",
                "id": "d111da240cd9ab17",
                "adTrackingEnabled": true,
                "manufacturer": "samsung"
            },
            "os": {
                "version": "4.4.4",
                "name": "REL",
                "sdk": 19
            },
            "screen": {
                "densityBucket": "xxhdpi",
                "density": 3,
                "height": 1920,
                "width": 1080,
                "densityDpi": 480,
                "scaledDensity": 3
            }
        }
    },
    "ip": "",
    "event": "{\"code\": \"mobile\", \"new_time\": 104.268, \"old_time\": 134.268, \"type\": \"onSkipSeek\", \"id\": \"i4x-MITx-15_390_1x-video-174238d48fd547269b05a7d0dcaa8394\", \"requested_skip_interval\": -30}",
    "event_type": "seek_video"}
}
{
    "username": "staff",
    "event_source": "mobile",
    "name": "edx.video.position.changed",
    "accept_language": "",
    "time": "2015-03-04T18:40:26.553000+00:00",
    "agent": "Dalvik/1.6.0 (Linux; U; Android 4.4.4; SM-G900V Build/KTU84P)",
    "page": "http://mobile3.m.sandbox.edx.org/courses/MITx/15.390.1x/3T2014/courseware/1a449d2fa9c8414697d550e53f387d0c/9d0771e62a3f457bb6d5ca131bc6f921",
    "host": "mobile3.m.sandbox.edx.org",
    "session": "",
    "referer": "",
    "context": {
        "component": "videoplayer",
        "received_at": "2015-03-04T18:40:26.553000+00:00",
        "course_id": "MITx/15.390.1x/3T2014",
        "path": "/segmentio/event",
        "user_id": 4,
        "org_id": "MITx",
        "application": {
            "version": "1.0.04",
            "name": "edx.mobileapp.android"
        },
        "client": {
            "network": {
                "wifi": true,
                "carrier": "Verizon Wireless",
                "cellular": true,
                "bluetooth": false
            },
            "locale": "en-US",
            "app": {
                "name": "edX",
                "packageName": "org.edx.mobile",
                "version": "1.0.04",
                "build": "org.edx.mobile@46",
                "versionName": "1.0.04",
                "versionCode": 46
            },
            "library": {
                "version": 247,
                "name": "analytics-android",
                "versionName": "2.4.7-SNAPSHOT"
            },
            "timezone": "America/New_York",
            "device": {
                "name": "kltevzw",
                "advertisingId": "d38b1986-5346-4321-994c-e0700ab76c39",
                "model": "SM-G900V",
                "type": "android",
                "id": "d111da240cd9ab17",
                "adTrackingEnabled": true,
                "manufacturer": "samsung"
            },
            "os": {
                "version": "4.4.4",
                "name": "REL",
                "sdk": 19
            },
            "screen": {
                "densityBucket": "xxhdpi",
                "density": 3,
                "height": 1920,
                "width": 1080,
                "densityDpi": 480,
                "scaledDensity": 3
            }
        }
    },
    "ip": "",
    "event": "{\"code\": \"mobile\", \"new_time\": 198.247, \"old_time\": 104.571, \"type\": \"onSlideSeek\", \"id\": \"i4x-MITx-15_390_1x-video-174238d48fd547269b05a7d0dcaa8394\", \"requested_skip_interval\": 93.676}",
    "event_type": "seek_video"
}
```

BenjiLee commented:
@lamagnifica

lamagnifica commented:
@BenjiLee the examples are super helpful, thanks. Doc PR to come soon.

BenjiLee commented:
@nasthagiri Another review please.

BenjiLee commented:
@brianhw 

I have verified that iOS v1.0.1, 1.0.02, and Android v1.0.04 seem to be processed by the shim correctly. In the future, I would like to have separate unit tests per build so that we do not have to manually check all of these conditions/ cases.

BenjiLee commented:
please test this

BenjiLee commented:
test this please

nasthagiri commented on common/djangoapps/track/shim.py:
```diff
@@ -119,15 +121,40 @@ def __call__(self, event):
 
             del payload['module_id']
 
+        if 'context' not in event:
```
was this check needed for a certain app version?  Or did you just add this as defensive coding?

nasthagiri replied:
nvm.  I see it was moved from below.

nasthagiri commented on common/djangoapps/track/shim.py:
```diff
@@ -119,15 +121,40 @@ def __call__(self, event):
+
+        #For the Android build that isn't distinguishing between skip and seek
+        if 'requested_skip_interval' in payload:
+            if payload['requested_skip_interval'] != -30:
```
can we compare with both -30 and +30?  or take the absolute value of the requested_skip_interval?
this way, this code will handle both rewind and forward features.
or, just in case the android build ever sends +30 by mistake again (argh!), at least we won't change the seek_type.  :)

nasthagiri commented on common/djangoapps/track/shim.py:
```diff
@@ -119,15 +121,40 @@ def __call__(self, event):
+                    payload['seek_type'] = 'slide'
+
+        #For converting seek_type to seek and skip|slide to onSlideSeek|onSkipSeek
+        if 'seek_type' in payload:
```
As we discussed, suggestion to mode this block of code above the Android block above (and you'll need to replace 'seek_type' with 'OnSlideSleek' in the Android block).

brianhw commented on common/djangoapps/track/shim.py:
```diff
@@ -101,6 +103,10 @@ def __call__(self, event):
         if name not in NAME_TO_EVENT_TYPE_MAP:
             return
 
+        #Convert edx.video.seeked to edx.video.positiion.changed
```
nit:  s/positiion/position

brianhw replied:
Also, comments should have a space after the '#'.  I'm surprised that wasn't complained about by pep8/pylint.

brianhw commented:
Talked with @shnayder.  While adding build-specific conditionals in the shim is rather ugly, the understanding is that this will "fix" events coming from the existing builds out in the field from the date this shim goes live.  This is primarily because these events were not "official" yet.  In future we may not choose to fix things in the shim this way.  

But a quick logic question:  it seems that the value of '30' is special here, and that doing some sort of skip operation will make it move backwards 30 seconds in the video.  Is this not available within the first 30 seconds of the video?  The logic seems to assert that any value other than 30 is a slide event instead of a skip event.

BenjiLee commented:
@brianhw The skip back 30 seconds will always return a "requested_time_interval" of -30 regardless of the position E.G. skipping back thirty 15s into the video will still return -30. This logic does assert that any other value than 30 is a slide.

nasthagiri commented on common/djangoapps/track/views/tests/test_segmentio.py:
```diff
@@ -404,3 +405,207 @@ def test_video_event(self, name, event_type):
+        (-30, "onSkipSeek")
+    )
+    @unpack
+    def test_old_android_build(self, requested_skip_interval, type):
```
should rename 'type' to 'expected_type'.  I'm surprised this wasn't a pylint warning.

nasthagiri commented on common/djangoapps/track/views/tests/test_segmentio.py:
```diff
@@ -404,3 +405,207 @@ def test_video_event(self, name, event_type):
+        (1, "onSlideSeek"),
+        (-2, "onSlideSeek"),
+        (30, "onSkipSeek"),
+        (-30, "onSkipSeek")
```
This doesn't test the renaming of `edx.video.seeked` to `edx.video.position.changed`.  This change was needed for android as well.
Suggestion: Why not combine both of these test methods into one?  You can pass the application name and version as ddt parameters.  The end comparison of the expected values are the same anyway, right?

BenjiLee replied:
I added this test to the build it was effected by which is the iOS build below.

nasthagiri commented on common/djangoapps/track/shim.py:
```diff
@@ -119,16 +124,43 @@ def __call__(self, event):
-        if 'current_time' in payload:
-            payload['currentTime'] = payload.pop('current_time')
-
-        event['event'] = json.dumps(payload)
```
Why was this code moved down?  This will change the behavior of other events.

nasthagiri commented on common/djangoapps/track/shim.py:
```diff
@@ -122,13 +127,38 @@ def __call__(self, event):
 
-        if 'open_in_browser_url' in context:
-            page, _sep, _tail = context.pop('open_in_browser_url').rpartition('/')
-            event['page'] = page
```
So before this PR, the 'page' field modification here was not included as part the `json.dumps`.  I hope this won't cause any issues.  I wouldn't think so - but what do I know about what our data packages need?

nasthagiri commented on common/djangoapps/track/views/tests/test_segmentio.py:
```diff
@@ -404,3 +405,119 @@ def test_video_event(self, name, event_type):
+        """
+        Test backwards compatibility of previous app builds
+
+        iOS version 1.0.02 incorrect emits the skip back 30 seconds as +30
```
s/incorrect/incorrectly

nasthagiri commented on common/djangoapps/track/views/tests/test_segmentio.py:
```diff
@@ -404,3 +405,119 @@ def test_video_event(self, name, event_type):
+        iOS version 1.0.02 incorrect emits the skip back 30 seconds as +30
+        instead of -30.
+        Android version 1.0.02 Skip and slide were both being returned as a
+        skip. Skip or slide is determined by checking if the skip time is == -30
```
Also, to add in the docstring: edx.video.seeked was sent as the seek name instead of edx.video.position.changed.

nasthagiri commented on common/djangoapps/track/views/tests/test_segmentio.py:
```diff
@@ -404,3 +405,119 @@ def test_video_event(self, name, event_type):
+        actual_event = dict(self.get_event())
+        payload = json.loads(actual_event.pop('event'))
+        print payload
+        print expected_payload
```
please remove these print statements.

nasthagiri commented on common/djangoapps/track/views/tests/test_segmentio.py:
```diff
@@ -404,3 +405,119 @@ def test_video_event(self, name, event_type):
+        (1, 1, "type", "onSlideSeek", "onSlideSeek", "edx.video.seeked", "edx.video.position.changed", 'edx.mobileapp.android', '1.0.02'),
+        (-2, -2, "type", "onSlideSeek", "onSlideSeek", "edx.video.seeked", "edx.video.position.changed", 'edx.mobileapp.android', '1.0.02'),
+        (30, 30, "type", "onSkipSeek", "onSkipSeek", "edx.video.position.changed", "edx.video.position.changed",'edx.mobileapp.android', '1.0.02'),
+        (-30, -30, "type", "onSkipSeek", "onSkipSeek", "edx.video.position.changed", "edx.video.position.changed", 'edx.mobileapp.android', '1.0.02')
```
Nitpick: if possible, if you could add a brief comment about what each of these ddt's is testing.

nasthagiri commented:
:+1: looks good.  awesome work and patience!

BenjiLee commented:
moved to https://github.com/edx/edx-platform/pull/7275
